### ISSUE TITLE
`LightGCN` example

### ISSUE THREAD

**Issue by rusty1s** [pyg-team/pytorch_geometric#4182]
### 🚀 The feature, motivation and pitch

While we provide the `LightGCN` model, there is an example missing to showcase its use.

**References:**
* https://github.com/pyg-team/pytorch_geometric/discussions/3992
* [`LightGCN` with PyTorch Geometric blogpost by CS224W](https://medium.com/stanford-cs224w/lightgcn-with-pytorch-geometric-91bab836471e)

@nishithshowri006 Let me know if you have still interest in contributing this feature :)

**Comment by nishithshowri006** [comment]
Hey @rusty1s I found one more important thing in the PYG model is the due to this line 
`self.reset_parameters()`

The Xavier uniform is not working, can you please check the performance and confirm it, for me I am facing the issue. When I comment it out run it works pretty well and initialized the embeddings above.

**Comment by nishithshowri006** [comment]
So can you comment out the reset parameters here and initialize the embeddings after defining the embeddings. Because I too couldn't understand why it isn't working when initialized from inside of that method.
![image](https://user-images.githubusercontent.com/58651995/156340692-d2ef96d0-dae7-45d4-886d-8ede5075ae71.png)

**Comment by rusty1s** [comment]
Why would we want to do that? We still want to provide the functionality to reset parameters IMO. What changes when we leave it that way? The initialization will be indeed different, but this shouldn't really matter, right?

**Comment by nishithshowri006** [comment]
Yeah that's right..let me check somethings and I will share a colab book about some trouble I am facing.

**Comment by nishithshowri006** [comment]
Hey @rusty1s can you please run these notebooks I made performance plots of different models. [colab](https://colab.research.google.com/drive/1OOdlf9Nuu5zpWjK7tikW7dHp04yQSwyA?usp=sharing)

**Comment by rusty1s** [comment]
Thanks for sharing. We confirmed that the output of custom LightGCN and PyG's model is the same given the same initialization/model parameters, right? Shall we confirm again? Does the performance still hold when setting a different random seed?

**Comment by nishithshowri006** [comment]
I am extremely sorry, I was dumb as why pyg library version isn't working...the library needs to get updated right, I wasn't thinking about that and was confused why it wasn't working after pushing changes to the main repo. I am really sorry for wasting your time it was my bad

**Comment by rusty1s** [comment]
No worries. Glad that you identified the issue.

**Comment by jasperhyp** [comment]
This might be a dumb comment, but I observed that in the CS224W Blog post only one direction (user -> item) of the edges is present.  Since message propagation is also unidirectional, isn't the update of user/item embeddings imbalanced?

**Comment by jasperhyp** [comment]
> This might be a dumb comment, but I observed that in the CS224W Blog post only one direction (user -> item) of the edges is present. Since message propagation is also unidirectional, isn't the update of user/item embeddings imbalanced?

I checked discussions #3708 #3043 and found that it is suggested that an explicitly transposed adj mat be used.  I also noticed that in the original `LightGCN` implementation (with Pytorch), they explicitly built a (M+N)x(M+N) adj mat rather than the MxN one we are using in PyG:

```
adj_mat = sp.dok_matrix((self.n_users + self.m_items, self.n_users + self.m_items), dtype=np.float32)
adj_mat = adj_mat.tolil()
R = self.UserItemNet.tolil()
adj_mat[:self.n_users, self.n_users:] = R
adj_mat[self.n_users:, :self.n_users] = R.T
adj_mat = adj_mat.todok()

```

**Comment by jasperhyp** [comment]
Oh I just noticed in @nishithshowri006 your example, you included this line:

```python
train_edges = torch.cat([torch.tensor(train.iloc[:,[4,5]].values, dtype = torch.long),
                         torch.tensor(train.iloc[:,[5,4]].values, dtype = torch.long)
                         ],axis = 0)
```

which clearly follows the (M+N)x(M+N) pipeline :)

**Comment by jasperhyp** [comment]
Another issue if the blog post seems to be that during evaluation the embeddings are derived by feeding val edges into the model.  However, in transductive scenario only the training edges should be fed into the model at this time.

**Comment by rusty1s** [comment]
Yes, we are also following the `(M+N)x(M+N)` pattern although this is just an implementation detail. From a computational and memory perspective, this does not change anything (since both are sparse matrices). Can you point me to the issue with feeding val edges into the model?

**Comment by rusty1s** [comment]
I see. I haven't looked into the blog post that closely TBH. Our implementation is independent from it :)

Regarding your problem. You can convert your extra information independently for both users and items into a shared embedding space, and then concatenate it with the initial embedding matrix. Would that work for you? Otherwise, I am not sure what you mean with two "masked edge lists".

**Comment by rusty1s** [comment]
Related discussion: https://github.com/pyg-team/pytorch_geometric/discussions/6786 @SourabhRande

**Comment by wukuoo** [comment]
@rusty1s @nishithshowri006 Thanks for your work on the LightGCN example. Is there any further update on it? I'm trying to learn through example but found the link @nishithshowri006  shared is expired [colab](https://colab.research.google.com/drive/1OOdlf9Nuu5zpWjK7tikW7dHp04yQSwyA?usp=sharing).

**Comment by rusty1s** [comment]
We should have an example ready pretty soon. It's part of our current community sprint.

**Comment by sorushsaghari** [comment]
Hi. beacuse i really need this example i wanted to ask if there is any updates on this issue. 
thanks for your answer

**Comment by rusty1s** [comment]
@happykygo

**Comment by happykygo** [comment]
I am working on it and will have a PR next week.

**Comment by daviddavo** [comment]
The PR was merged and the example is available [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/lightgcn.py), so this issue can be closed

**Comment by sorushsaghari** [comment]
the dataset imported in the example is not accessible in torch_geometrics:
```python
from torch_geometric.datasets import AmazonBook
```
error:
```
ImportError: cannot import name 'AmazonBook' from 'torch_geometric.datasets' (/home/user_name/.local/lib/python3.10/site-packages/torch_geometric/datasets/__init__.py)
```

**Comment by akihironitta** [comment]
@sorushsaghari It hasn't been released yet and will be included in the next release, 2.4.0. In the meantime, you can use it by installing nightly or master. https://github.com/pyg-team/pytorch_geometric#nightly-and-master

**Comment by sorushsaghari** [comment]
> @sorushsaghari It hasn't been released yet and will be included in the next release, 2.4.0. In the meantime, you can use it by installing nightly or master. https://github.com/pyg-team/pytorch_geometric#nightly-and-master

ah, thanks a lot.
